.tools/phpstan/RexFunctionsDynamicReturnTypeExtension.php

```diff
@@ -0,0 +1,122 @@
+<?php
+
+declare(strict_types=1);
+
+namespace redaxo\phpstan;
+
+use PhpParser\Node\Expr\FuncCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\FunctionReflection;
+use PHPStan\Reflection\ParametersAcceptorSelector;
+use PHPStan\ShouldNotHappenException;
+use PHPStan\Type\ArrayType;
+use PHPStan\Type\BooleanType;
+use PHPStan\Type\Constant\ConstantStringType;
+use PHPStan\Type\DynamicFunctionReturnTypeExtension;
+use PHPStan\Type\FloatType;
+use PHPStan\Type\IntegerType;
+use PHPStan\Type\MixedType;
+use PHPStan\Type\ObjectWithoutClassType;
+use PHPStan\Type\StringType;
+use PHPStan\Type\Type;
+use PHPStan\Type\TypeCombinator;
+use PHPStan\Type\UnionType;
+use function count;
+use function in_array;
+
+final class RexFunctionsDynamicReturnTypeExtension implements DynamicFunctionReturnTypeExtension
+{
+    public function isFunctionSupported(FunctionReflection $functionReflection): bool
+    {
+        return in_array($functionReflection->getName(), ['rex_get', 'rex_post', 'rex_request', 'rex_server', 'rex_session', 'rex_cookie', 'rex_files', 'rex_env'], true);
+    }
+
+    public function getTypeFromFunctionCall(
+        FunctionReflection $functionReflection,
+        FuncCall $functionCall,
+        Scope $scope
+    ): Type {
+        $args = $functionCall->getArgs();
+
+        if (count($args) < 2) {
+            return ParametersAcceptorSelector::selectSingle($functionReflection->getVariants())->getReturnType();
+        }
+
+        $defaultArgType = null;
+        if (count($args) >= 3) {
+            $defaultArgType = $scope->getType($args[2]->value);
+        }
+
+        $typeString = $scope->getType($args[1]->value);
+        if ($typeString instanceof ConstantStringType) {
+            $resolvedType = $this->resolveTypeFromString($typeString->getValue());
+
+            if (null !== $resolvedType) {
+                if ($defaultArgType) {
+                    return TypeCombinator::union($resolvedType, $defaultArgType);
+                }
+                return $resolvedType;
+            }
+        }
+
+        return ParametersAcceptorSelector::selectSingle($functionReflection->getVariants())->getReturnType();
+    }
+
+    private function resolveTypeFromString(string $vartype): ?Type
```

Comment:
@gharlan könntest du hier mal drüber schauen was hier noch fehlt, damit diese extension hier "genauso schlau" ist wie die im psalm-plugin?

Reply:
Wenn ich das richtig sehe, sind vor allem die Typen in Array-Struktur nicht implementiert. Also [sowas](https://github.com/redaxo/redaxo/blob/c9e667d630870942dfa5568bf7be56c8a89661db/redaxo/src/addons/phpmailer/pages/config.php#L21-L43).